internal/wsutil/wsutil.go

```diff
@@ -109,48 +81,61 @@ func WebsocketSendStream(conn MessageWriter, r io.Reader, bufferSize int) chan b
 func WebsocketRecvStream(w io.Writer, conn MessageReader) chan bool {
 	ch := make(chan bool)
 
-	go func(w io.Writer, conn MessageReader) {
-		for {
-			mt, r, err := conn.NextReader()
-			if mt == websocket.CloseMessage {
-				logger.Debugf("Got close message for reader")
-				break
-			}
-
-			if mt == websocket.TextMessage {
-				logger.Debugf("Got message barrier")
-				break
-			}
+	go func() {
+		recvLoop(w, conn)
+		close(ch)
+	}()
 
-			if err != nil {
-				logger.Debugf("Got error getting next reader %s", err)
-				break
-			}
+	return ch
+}
 
-			buf, err := ioutil.ReadAll(r)
-			if err != nil {
-				logger.Debugf("Got error writing to writer %s", err)
-				break
-			}
+func recvLoop(w io.Writer, conn MessageReader) {
+	for {
+		mt, r, err := conn.NextReader()
+		if err != nil {
+			logger.Debugf("Cannot get next reader: %v", err)
+			return
+		}
 
-			if w == nil {
-				continue
-			}
+		if mt == websocket.CloseMessage {
+			logger.Debugf("Got close message for reader")
+			return
+		}
 
-			i, err := w.Write(buf)
-			if i != len(buf) {
-				logger.Debugf("Didn't write all of buf")
-				break
+		if mt == websocket.TextMessage {
+			var command struct {
+				Command string `json:"command"`
 			}
+			err := json.NewDecoder(r).Decode(&command)
 			if err != nil {
-				logger.Debugf("Error writing buf %s", err)
-				break
+				logger.Noticef("Cannot decode I/O command: %v", err)
+				continue
+			}
+			switch command.Command {
+			case "end":
+				logger.Debugf(`Got message barrier ("end" command)`)
+				return
+			default:
+				logger.Noticef("Invalid I/O command %q", command.Command)
+				continue
 			}
 		}
-		ch <- true
-	}(w, conn)
 
-	return ch
+		buf, err := ioutil.ReadAll(r)
+		if err != nil {
+			logger.Debugf("Cannot read from message reader: %v", err)
+			return
```

Comment:
Should these be more visible errors somehow?  What happens elsewhere if we return silently here and below?

Reply:
I started doing that and a couple of them do actually occur under normal conditions, for example error getting the reader if one side closes the connection first. So I decided to just leave for now. Can always go into `PEBBLE_DEBUG=1` mode if/when we need to debug.